### PR TITLE
Spectrum Annotations front end

### PR DESCRIPTION
**Pull request by guynir42**
Adding the annotations table from the source page / scanning page, but for each individual spectrum in the ManageData page. 
Screenshot from 2021-10-27 14-42-34

### CODE REVIEW

**Comment by acrellin** [comment]
> I have no idea how to make the three components in the pull out row behave (i.e., sit in one line). Any help would be appreciated! @acrellin

@guynir42 What's the issue exactly? Happy to try to help, just don't know what you want to do here.

File: static/js/components/SourceDesktop.jsx [hunk 1]
```diff
@@ -484,7 +484,10 @@ const SourceDesktop = ({ source }) => {
               </Typography>
             </AccordionSummary>
             <AccordionDetails>
-              <ObjPageAnnotations annotations={source.annotations} />
+              <ObjPageAnnotations
+                annotations={source.annotations}
+                spectra={spectra}
```
Comment (acrellin) [inline static/js/components/SourceDesktop.jsx:489]: Why not just pass in spectrum annotations after they've been processed rather than the entire spectra state?

File: skyportal/handlers/api/spectrum.py [hunk 2]
```diff
@@ -635,6 +653,10 @@ def get(self, obj_id):
                 key=lambda x: x["created_at"],
                 reverse=True,
             )
+            annotations = [
+                {**a.to_dict(), 'author': a.author.to_dict()} for a in annotations
```
Comment (acrellin) [inline skyportal/handlers/api/spectrum.py:657]: Why is this necessary?

File: skyportal/handlers/api/spectrum.py [hunk 3]
```diff
@@ -195,6 +196,14 @@ def get(self, spectrum_id):
             .filter(CommentOnSpectrum.spectrum_id == spectrum_id)
             .all()
         )
+        annotations = (
+            AnnotationOnSpectrum.query_records_accessible_by(
+                self.current_user,
+                options=[joinedload(AnnotationOnSpectrum.groups)],
```
Comment (acrellin) [inline skyportal/handlers/api/spectrum.py:202]: Does the front-end even utilize the groups list for each annotation? If not, let's omit this

File: skyportal/handlers/api/spectrum.py [hunk 4]
```diff
@@ -616,6 +626,14 @@ def get(self, obj_id):
                 .filter(CommentOnSpectrum.spectrum_id == spec.id)
                 .all()
             )
+            annotations = (
+                AnnotationOnSpectrum.query_records_accessible_by(
+                    self.current_user,
+                    options=[joinedload(AnnotationOnSpectrum.groups)],
```
Comment (acrellin) [inline skyportal/handlers/api/spectrum.py:632]: Same question here

**Comment by guynir42** [comment]
> > I have no idea how to make the three components in the pull out row behave (i.e., sit in one line). Any help would be appreciated! @acrellin
> 
> @guynir42 What's the issue exactly? Happy to try to help, just don't know what you want to do here.

If you look at the example image above (or if you check out this branch) you'll see that the comments component is stretched quite a bit horizontally, and then the annotations component is dropped to the second row. I think it would look better if I could get them more compactly into one row (at least in cases where the browser width is large enough). 
I guess if I could coerce each component to have a maximum size that would do it (maybe?) but I'm not sure where to control that.

File: skyportal/plot.py [hunk 5]
```diff
@@ -1380,7 +1381,9 @@ def spectroscopy_plot(obj_id, user, spec_id=None, width=600, device="browser"):
 
     obj = Obj.get_if_accessible_by(obj_id, user)
     spectra = (
-        Spectrum.query_records_accessible_by(user)
+        Spectrum.query_records_accessible_by(
+            user, options=[joinedload(Spectrum.annotations)]
```
Comment (acrellin) [inline skyportal/plot.py:1385]: This is not right -- you'll need to use a subquery here, becase annotations on spectra can have different permissioning than their associated spectrum (access to annotations on spectra is determined by group membership, not only whether they can read the underlying spectrum).

File: skyportal/plot.py [hunk 6]
```diff
@@ -1409,6 +1412,12 @@ def spectroscopy_plot(obj_id, user, spec_id=None, width=600, device="browser"):
         normfac = np.nanmedian(np.abs(s.fluxes))
         normfac = normfac if normfac != 0.0 else 1e-20
         altdata = json.dumps(s.altdata) if s.altdata is not None else ""
+        annotations = (
+            json.dumps([{a.origin: a.data} for a in s.annotations])
```
Comment (acrellin) [inline skyportal/plot.py:1416]: Why is this necessary?

File: static/js/components/ManageDataForm.jsx [hunk 7]
```diff
@@ -136,13 +137,21 @@ const useStyles = makeStyles(() => ({
   },
 }));
 
-const SpectrumRow = ({ rowData, route }) => {
+const SpectrumRow = ({ rowData, route, annotations }) => {
   const styles = useSourceStyles();
   const colSpan = rowData.length + 1;
+  const spectrumID =
+    typeof rowData[0] === "number" ? rowData[0] : parseInt(rowData[0], 10);
```
Comment (acrellin) [inline static/js/components/ManageDataForm.jsx:144]:
```suggestion
  const spectrumID = parseInt(rowData[0], 10);
```
no harm in passing in a value that's already an int; this is much easier to read/maintain.

File: static/js/components/ManageDataForm.jsx [hunk 8]
```diff
@@ -161,13 +170,13 @@ const SpectrumRow = ({ rowData, route }) => {
...
-                typeof rowData[0] === "number"
-                  ? rowData[0]
-                  : parseInt(rowData[0], 10)
-              }
+              spectrumID={spectrumID}
             />
           </Grid>
+          <Grid item sm={6}>
+            <Typography variant="h6">Annotations</Typography>
+            <ObjPageAnnotations annotations={annotations} />
```
Comment (acrellin) [inline static/js/components/ManageDataForm.jsx:178]: Now that this component is being used in places besides the Obj Page, it makes sense to rename it to `ObjAnnotations`

File: static/js/components/ObjPageAnnotations.jsx [hunk 9]
```diff
@@ -69,11 +69,21 @@ const getMuiTheme = (theme) =>
   });
 
 // Table for displaying Obj annotations on Candidate page and Source page
-const ObjPageAnnotations = ({ annotations }) => {
+const ObjPageAnnotations = ({ annotations, spectra = [] }) => {
```
Comment (acrellin) [inline static/js/components/ObjPageAnnotations.jsx:72]: See comment about passing in spectrum annotations, not the entire spectra list

File: static/js/components/ObjPageAnnotations.jsx [hunk 10]
```diff
@@ -69,11 +69,21 @@ const getMuiTheme = (theme) =>
...
   const classes = useStyles();
   const theme = useTheme();
   const renderValue = (value) => getAnnotationValueString(value);
   const renderTime = (created_at) => dayjs().to(dayjs.utc(`${created_at}Z`));
+  const renderSpectrumDate = (observed_at) => {
+    if (observed_at) {
+      const dayFraction = (parseFloat(observed_at.substring(11, 13)) / 24) * 10;
+      return [observed_at.substring(0, 10), ".", dayFraction.toFixed(0)].join(
+        ""
+      );
```
Comment (acrellin) [inline static/js/components/ObjPageAnnotations.jsx:82]:
```suggestion
      return observed_at.substring(0, 10) + "." + dayFraction.toFixed(0);
```

Comment (acrellin) [reply static/js/components/ObjPageAnnotations.jsx:82]: Simpler and easier to maintain

Comment (acrellin) [reply static/js/components/ObjPageAnnotations.jsx:82]: or perhaps preferably:
``` javascript
return `${observed_at.substring(0, 10)}.${dayFraction.toFixed(0)}`;
```

File: static/js/components/ObjPageAnnotations.jsx [hunk 11]
```diff
@@ -112,6 +122,38 @@ const ObjPageAnnotations = ({ annotations }) => {
     },
   ];
 
+  // if there are any spectra given, we should add the spectrum annotations after object annotations
+  if (spectra.length > 0) {
```
Comment (acrellin) [inline static/js/components/ObjPageAnnotations.jsx:126]: Is it possible for `spectra` to ever be passed in as `null` or `undefined` while the data are still being fetched? If so we need to be able to handle that here gracefully (this will error as-is if so).

File: static/js/components/ObjPageAnnotations.jsx [hunk 12]
```diff
@@ -69,11 +69,21 @@ const getMuiTheme = (theme) =>
   });
 
 // Table for displaying Obj annotations on Candidate page and Source page
```
Comment (acrellin) [inline static/js/components/ObjPageAnnotations.jsx:71]:
```suggestion
// Table for displaying Obj annotations
```

File: static/js/components/ObjPageAnnotations.jsx [hunk 13]
```diff
@@ -112,6 +122,38 @@ const ObjPageAnnotations = ({ annotations }) => {
     },
   ];
 
+  // if there are any spectra given, we should add the spectrum annotations after object annotations
+  if (spectra.length > 0) {
+    // first add the observed_at property to each annotation in tableData:
+    tableData.forEach((ann) => {
```
Comment (acrellin) [inline static/js/components/ObjPageAnnotations.jsx:128]: Let's do this above where the regular obj annotations are added with a conditional check.

**Review by acrellin — CHANGES_REQUESTED** [review]
Looking good so far! In addition to the below comments, this needs tests.

File: static/js/components/SourceDesktop.jsx (hunk 1 above)
Comment (guynir42) [reply static/js/components/SourceDesktop.jsx:489]: The annotations require additional info like the time each spectrum was detected. That means you have to start passing around data from spectra to annotations before giving it to the component. 
It seemed cleaner to just give the component this one optional argument.

File: skyportal/handlers/api/spectrum.py (hunk 2 above)
Comment (guynir42) [reply skyportal/handlers/api/spectrum.py:657]: I think without this the annotations do not come with an author (which is displayed in the table).

File: skyportal/handlers/api/spectrum.py (hunk 3 above)
Comment (guynir42) [reply skyportal/handlers/api/spectrum.py:202]: Maybe you are right...

File: skyportal/plot.py (hunk 5 above)
Comment (guynir42) [reply skyportal/plot.py:1385]: OK so query spectrum annotations (accessible by) and then filter on spectrum ID. Got it.

File: skyportal/plot.py (hunk 6 above)
Comment (guynir42) [reply skyportal/plot.py:1416]: Actually I'm not sure how the data would be rendered without this. I can give it a try.

File: static/js/components/ManageDataForm.jsx (hunk 8 above)
Comment (guynir42) [reply static/js/components/ManageDataForm.jsx:178]: I think it should be named `AnnotationsTable` since it also appears on the ManageData page.

File: static/js/components/ObjPageAnnotations.jsx (hunk 10 above)
Comment (guynir42) [reply static/js/components/ObjPageAnnotations.jsx:82]: I will go with option 2 because option 1 is what I did but the Lint didn't like that.

File: static/js/components/ObjPageAnnotations.jsx (hunk 13 above)
Comment (guynir42) [reply static/js/components/ObjPageAnnotations.jsx:128]: Not sure what you mean?

File: skyportal/handlers/api/spectrum.py (hunk 2 above)
Comment (guynir42) [reply skyportal/handlers/api/spectrum.py:657]: Nevermind it works without this. I don't know why but for comments this was needed (maybe there it required the gravator or something).

File: skyportal/plot.py (hunk 6 above)
Comment (guynir42) [reply skyportal/plot.py:1416]: I think the json dump is needed to turn the objects into strings that bokeh can render.

File: skyportal/handlers/api/spectrum.py (hunk 2 above)
Comment (guynir42) [reply skyportal/handlers/api/spectrum.py:657]: I don't know why we need this here and not in the object annotations. But it doesn't render an author if I don't add this...

**Comment by profjsb** [comment]
Apart from tests and the remaining enquires from @acrellin, this is looking good.

File: static/js/components/SourceDesktop.jsx (hunk 1 above)
Comment (acrellin) [reply static/js/components/SourceDesktop.jsx:489]: Well, for consistency, I think we should do the pre-processing before we pass the data in here so that we have consistent props like `annotations={...} specAnnotations = {...}`

File: skyportal/handlers/api/spectrum.py (hunk 2 above)
Comment (acrellin) [reply skyportal/handlers/api/spectrum.py:651]: Okay, this is okay to leave as is then. Thanks

File: skyportal/plot.py (hunk 6 above)
Comment (acrellin) [reply skyportal/plot.py]: Okay

File: static/js/components/ObjPageAnnotations.jsx (hunk 11 above)
Comment (acrellin) [reply static/js/components/ObjPageAnnotations.jsx:124]: Still wondering this

File: static/js/components/ObjPageAnnotations.jsx (hunk 13 above)
Comment (acrellin) [reply static/js/components/ObjPageAnnotations.jsx:126]: Above, there's an iteration over each annotation, where we should add this field if relevant there so we're not iterating again unnecessarily

File: static/js/components/ObjPageAnnotations.jsx (hunk 11 above)
Comment (guynir42) [reply static/js/components/ObjPageAnnotations.jsx:124]: Yeah sorry I think we can just add a couple more conditions to that if, that's not a problem.

Comment (acrellin) [reply static/js/components/ObjPageAnnotations.jsx:124]: like `if (spectra && spectra.length)`

File: static/js/components/AnnotationsTable.jsx [hunk 14]
```diff
@@ -84,10 +84,13 @@ const AnnotationsTable = ({ annotations, spectra = [] }) => {
 
   // Curate data
   const tableData = [];
+  annotations.push(...spectrumAnnotations);
```
Comment (acrellin) [inline static/js/components/AnnotationsTable.jsx:87]: If `annotations` could ever possibly be `null`, we'd need to add `annotations?.push(...`

File: static/js/components/AnnotationsTable.jsx [hunk 15]
```diff
@@ -120,30 +123,7 @@ const AnnotationsTable = ({ annotations, spectra = [] }) => {
...
-            value,
-            author,
-            created_at,
-            observed_at: spec.observed_at,
-          });
-        });
-      });
-    });
-
+  if (spectrumAnnotations && spectrumAnnotations.length) {
```
Comment (acrellin) [inline static/js/components/AnnotationsTable.jsx:126]: `  if (spectrumAnnotations?.length) {`

File: static/js/components/AnnotationsTable.jsx [hunk 16]
```diff
@@ -84,10 +84,13 @@ const AnnotationsTable = ({ annotations, spectra = [] }) => {
 
   // Curate data
   const tableData = [];
+  annotations.push(...spectrumAnnotations);
   annotations?.forEach((annotation) => {
-    const { origin, data, author, created_at } = annotation;
+    const { origin, data, author, created_at, spectrum_observed_at } =
+      annotation;
+    const observed_at = spectrum_observed_at || null;
```
Comment (acrellin) [inline static/js/components/AnnotationsTable.jsx:91]:
``` javascript
    const { origin, data, author, created_at, spectrum_observed_at: observed_at = null } =
      annotation;
```

File: static/js/components/AnnotationsTable.jsx [hunk 17]
```diff
@@ -68,18 +68,29 @@ const getMuiTheme = (theme) =>
...
+  };
+
   // Curate data
   const tableData = [];
+  annotations.push(...spectrumAnnotations);
   annotations?.forEach((annotation) => {
-    const { origin, data, author, created_at } = annotation;
+    const { origin, data, author, created_at, spectrum_observed_at } =
+      annotation;
+    const observed_at = spectrum_observed_at || null;
```
Comment (acrellin) [inline static/js/components/AnnotationsTable.jsx:91]:
```suggestion
    const { origin, data, author, created_at, spectrum_observed_at: observed_at = null } =
      annotation;
```

**Comment by acrellin** [comment]
@guynir42 Looking much better, thanks! A few more suggestions above. And can you also add some spectrum annotations to the demo data yaml so that I can check out this branch and easily see it in action?

**Comment by guynir42** [comment]
> @guynir42 Looking much better, thanks! A few more suggestions above. And can you also add some spectrum annotations to the demo data yaml so that I can check out this branch and easily see it in action?

Good idea!

**Review by acrellin — CHANGES_REQUESTED** [review]
Looks good except that 14gqr and 16fil are in the process of being replaced; can you please update the demo data yaml to post to other sources? And add regular annotations there if necessary so we can see them alongside each other on the source page.

**Comment by guynir42** [comment]
> Looks good except that 14gqr and 16fil are in the process of being replaced; can you please update the demo data yaml to post to other sources? And add regular annotations there if necessary so we can see them alongside each other on the source page.

Done @acrellin

**Review by acrellin — APPROVED** [review]
Looks good, thanks @guynir42!

**Comment by profjsb** [comment]
@guynir42 can you resolve the conflicts with the demo data? After that, I can approve and merge.

**Review by profjsb — APPROVED** [review]
